server/s3_copy_object_action.cc

```diff
@@ -282,26 +274,294 @@ void S3CopyObjectAction::fetch_object_info_success() {
+  new_object_metadata.reset(new S3ObjectMetadata(*source_object_metadata));
+
+  new_object_metadata->set_object_list_index_oid(
+      bucket_metadata->get_object_list_index_oid());
+
+  new_object_metadata->set_objects_version_list_index_oid(
+      bucket_metadata->get_objects_version_list_index_oid());
+
+  new_oid_str = S3M0Uint128Helper::to_string(new_object_oid);
+
+  // Generate a version id for the new object.
+  new_object_metadata->regenerate_version_id();
+  new_object_metadata->set_oid(motr_writer->get_oid());
+  new_object_metadata->set_layout_id(layout_id);
+
   next();
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
 
+void S3CopyObjectAction::create_object_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  if (check_shutdown_and_rollback()) {
+    s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+    return;
+  }
+  if (motr_writer->get_state() == S3MotrWiterOpState::exists) {
+    // collision_detected();
+  } else {
+    s3_copy_action_state = S3CopyObjectActionState::newObjOidCreationFailed;
+
+    if (motr_writer->get_state() == S3MotrWiterOpState::failed_to_launch) {
+      s3_log(S3_LOG_ERROR, request_id, "Create object failed.\n");
+      set_s3_error("ServiceUnavailable");
+    } else {
+      s3_log(S3_LOG_WARN, request_id, "Create object failed.\n");
+      // Any other error report failure.
+      set_s3_error("InternalError");
+    }
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
 // Copy source object to destination object
 void S3CopyObjectAction::copy_object() {
   s3_log(S3_LOG_INFO, request_id, "Entering\n");
-  read_object();              // read source object
-  initiate_data_streaming();  // write to destination object
-  next();
+
+  if (!content_length) {
+    s3_log(S3_LOG_DEBUG, request_id, "Source object is empty");
+    next();
+    return;
+  }
+  motr_reader = motr_reader_factory->create_motr_reader(
+      request, source_object_metadata->get_oid(),
+      source_object_metadata->get_layout_id());
+  motr_reader->set_last_index(0);
+
+  rest_bytes = content_length;
+  read_data_block();
+
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
 
-// Save destination object metadata
+void S3CopyObjectAction::read_data_block() {
+  s3_log(S3_LOG_INFO, request_id, "Entering");
+
+  assert(!read_in_progress);
+  assert(rest_bytes > 0);
+
+  const size_t max_blocks_in_one_read_op =
+      S3Option::get_instance()->get_motr_units_per_request();
+  const size_t rest_blocks = (rest_bytes + motr_unit_size - 1) / motr_unit_size;
+
+  if (motr_reader->read_object_data(
+          std::min(rest_blocks, max_blocks_in_one_read_op),
+          std::bind(&S3CopyObjectAction::read_data_block_success, this),
+          std::bind(&S3CopyObjectAction::read_data_block_failed, this))) {
+
+    s3_log(S3_LOG_DEBUG, request_id, "Data read is launched");
+    read_in_progress = true;
+    return;
+  }
+  if (motr_reader->get_state() == S3MotrReaderOpState::failed_to_launch) {
```

Comment:
should we add this section in read_data_block_failed() ?

Reply:
I borrowed this code snippet from s3_get_object.cc